### PR TITLE
Cadence CLI

### PR DESCRIPTION
**Pull request by vancexu**
command line tool for users to perform common tasks with cadence, see readme for details. 
Tests will be done in separate PR

### CODE REVIEW

File: common/cli/util.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		data = e.TimerCanceledEventAttributes
+
+	case s.EventTypeMarkerRecorded:
+		data = e.MarkerRecordedEventAttributes
+
+	case s.EventTypeWorkflowExecutionTerminated:
+		data = e.WorkflowExecutionTerminatedEventAttributes
+
+	default:
+		data = e
```
Comment (wxing1292) [inline common/cli/util.go]: this list of events is not complete, at least missing the child workflow events

File: common/cli/factory.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		return Staging
+	}
+	if env == "production" {
+		return Production
+	}
+	return Development
+}
+
+// WorkflowClientBuilderInterface is an interface to build client to cadence service
+type WorkflowClientBuilderInterface interface {
```
Comment (madhuravi) [inline common/cli/factory.go]: Why is this file here? Where is it being used? Can't find any references in the diff.

File: common/cli/util.go (hunk 1 above)
Comment (vancexu) [reply common/cli/util.go]: Good catch, added all

File: common/cli/factory.go (hunk 2 above)
Comment (vancexu) [reply common/cli/factory.go]: I previously thought common might be a good place, and later playground or other apps can use it. But now I think it might be better just put everything in cli folder for now, so changed. 
This interface is used in "commands.go", and I need it so that I can have a customized builder used internal. See update comments in the file.

File: tools/cli/app.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+// NewCliApp instantiates a new instance of the CLI application.
+func NewCliApp() *cli.App {
+	app := cli.NewApp()
+	app.Name = "cadence"
+	app.Usage = "A command-line tool for cadence users"
+	app.Version = Version
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:   FlagAddressWithAlias,
+			Value:  "",
```
Comment (yiminc) [inline tools/cli/app.go]: I think we have default value of 127.0.0.1:7933, it is not the case?

File: tools/cli/app.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		},
+		cli.StringFlag{
+			Name:   FlagDomainWithAlias,
+			Usage:  "cadence workflow domain",
+			EnvVar: "CADENCE_CLI_DOMAIN",
+		},
+		cli.BoolFlag{
+			Name:   FlagCustomizeBuilderWithAlias,
+			Usage:  "customize builder function or default",
+			EnvVar: "CADENCE_CUSTOMIZED_BUILDER",
```
Comment (yiminc) [inline tools/cli/app.go]: this flag is confusing. How is user suppose to use this flag?

File: tools/cli/commands.go [hunk 5]
```diff
@@ -0,0 +1,871 @@
...
+	defaultDomainRetentionDays = 3
+)
+
+// For color output to terminal
+var colorRed func(i ...interface{}) string
+var colorMagenta func(i ...interface{}) string
+var colorGreen func(i ...interface{}) string
+
+func init() {
+	colorRed = color.New(color.FgRed).SprintFunc()
```
Comment (yiminc) [inline tools/cli/commands.go]: could you assign the value when you declare the variables so we don't need this init()?

File: tools/cli/commands.go [hunk 6]
```diff
@@ -0,0 +1,871 @@
...
+	colorMagenta = color.New(color.FgMagenta).SprintFunc()
+	colorGreen = color.New(color.FgGreen).SprintFunc()
+}
+
+// cBuilder is used to create cadence clients if user want to provide customized builder
+var cBuilder WorkflowClientBuilderInterface
```
Comment (yiminc) [inline tools/cli/commands.go]: I don't understand why do we need this?

File: tools/cli/commands.go [hunk 7]
```diff
@@ -0,0 +1,871 @@
...
+	}
+	return fmt.Sprintf("cadence-cli@%s", hostName)
+}
+
+func newContext() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(context.Background(), time.Second*5)
+}
+
+func newContextForLongPoll() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(context.Background(), time.Minute*2)
```
Comment (yiminc) [inline tools/cli/commands.go]: Do we need to let user specify timeout for this? (Make it optional, and default to 2 minutes.) For example, what if the workflow takes longer than 2 minutes?

File: tools/cli/app.go (hunk 3 above)
Comment (vancexu) [reply tools/cli/app.go]: Default `localHostPort` is 127.0.0.1:7933. I put empty here is easier for processing.

File: tools/cli/app.go (hunk 4 above)
Comment (vancexu) [reply tools/cli/app.go]: Agree it is confusing, removed.

File: tools/cli/commands.go (hunk 5 above)
Comment (vancexu) [reply tools/cli/commands.go]: changed

File: tools/cli/commands.go (hunk 6 above)
Comment (vancexu) [reply tools/cli/commands.go]: This is necessary for customized builder. Internally, we have special tag like ENV and additional processing when creating builder. By implementing these interfaces and inject with SetBuilder, we can achieve customized builder. This is transparent to CLI user.

File: tools/cli/commands.go (hunk 7 above)
Comment (vancexu) [reply tools/cli/commands.go]: Yes, changed.

File: tools/cli/README.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+./cadence-cli --domain samples-domain workflow start --tl helloWorldGroup --wt main.Workflow --et 60 --dt 10 -i '"vancexu"'
+
+# see more options for workflow start
+./cadence-cli --domain samples-domain workflow start -h
+
+# for workflow with multiple input, seperate each json with space/newline like
+./cadence-cli --domain samples-domain workflow start --tl helloWorldGroup --wt main.WorkflowWith3Args --et 60 --dt 10 -i '"your_input_string" 123 {"Name":"my-string", "Age":12345}'
+```
+
+- Run workflow and see the progress
```
Comment (mfateev) [inline tools/cli/README.md]: I think this one should come before start as most developers should use it during development. I think you also want to expand on what "run" means in this context and what benefits over start it has.

File: tools/cli/README.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+*Tips:*  
+to avoid repeated input global option **domain**, user can export domain-name in environment variable CADENCE_CLI_DOMAIN.
+(There are also other environment variables like this, see help message for details)
+```
+export CADENCE_CLI_DOMAIN=samples-domain
+
+# then you can just
+./cadence-cli domain desc
+./cadence-cli workflow list
+./cadence-cli workflow show -w greetings_38a9e481-a00c-433a-b1ff-b6f67536efd8
```
Comment (mfateev) [inline tools/cli/README.md]: Can we make -w optional after show? Just workflow show UUID?

File: tools/cli/README.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+Cadence CLI provide an command-line tool for users to perform various tasks on Cadence.  
+Users can perform operations like register, update and describe on domain;  
+also start workflow, show workflow history, signal workflow ... and many other tasks.    
+
+## How
+- Run `make bins`
+- You should see an executable `cadence-cli`
+
+## Quick Start
+Run `./cadence-cli` to view help message. There are some top level commands and global options.
```
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: Yes internally user use `cadence` through brew install, but the runnable we generate is `./cadence-cli`. In public server repo, `cadence` is used as the runnable to start cadence server

[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: Maybe I can combine currently `cadence` and `cadence-cli` to one, and create commands like `cadence service start`? What do you think?
Otherwise we have to use `cadence-cli ...`

[the remark this answers is not in the document]
Comment (mfateev) [reply tools/cli/README.md]: Let's rename cadence to cadence-server and cadence-cli to cadence then

File: tools/cli/README.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+## Quick Start
+Run `./cadence-cli` to view help message. There are some top level commands and global options.   
+Run `./cadence-cli domain` to view help message about operations on domain  
+Run `./cadence-cli workflow` to view help message about operations on workflow  
+
+*Note:* make sure you have cadence server running before using CLI 
+
+### Domain operation examples 
+- Register a new domain named "samples-domain":
```
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: Good suggestion, I will add DOMAIN env here also. 
File support will be done as a feature, together with cache workflow_id and run_id

File: tools/cli/README.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+### Domain operation examples 
+- Register a new domain named "samples-domain":  
+```
+./cadence-cli --domain samples-domain domain register 
+# OR using short alias  
+./cadence-cli --do samples-domain domain re
+```   
+- View "samples-domain" details:   
+```
+./cadence-cli --domain samples-domain domain describe
```
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: I think "workflow describe` is a little bit confusing which looks like describe detail info of workflow such as workflow type, arguments, wid, runid etc, but `workflow show` is showing history.

File: tools/cli/factory.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+const (
+	_cadenceClientName      = "cadence-client"
+	_cadenceFrontendService = "cadence-frontend"
+)
+
+// WorkflowClientBuilderInterface is an interface to build client to cadence service
+// Usually default WorkflowClientBuilder is enough for development purpose, but
+// user can still customized builder by implementing this interface, and call SetBulder after initialize cli.
+// The customized builder may have more processing on Env, Address and other info.
+type WorkflowClientBuilderInterface interface {
```
Comment (yiminc) [inline tools/cli/factory.go]: I thought we talked about this interface, and decided that it only need one method:
CreateServiceClient(ctx *cli.Context). Or am I missing something?

File: tools/cli/README.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+### Workflow operation examples
+
+- Start workflow: (for example [this workflow](https://github.com/samarabbas/cadence-samples/blob/master/cmd/samples/recipes/helloworld/helloworld_workflow.go))
+```
+./cadence-cli --domain samples-domain workflow start --tl helloWorldGroup --wt main.Workflow --et 60 --dt 10 -i '"vancexu"'
+
+# see more options for workflow start
+./cadence-cli --domain samples-domain workflow start -h
```
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: No, `cadence workflow start help` won't work because there is no `help` subsubcommands, but only `--help` or `-h` command options. 

When view help of `cadence workflow` it will show below:
```
NAME:
   cadence workflow - Operate cadence workflow

USAGE:
   cadence workflow command [command options] [arguments...]

COMMANDS:
     show             show workflow history
     start            start a new workflow execution
     run              start a new workflow execution and get workflow progress
     cancel, c        cancel a workflow execution
     signal, s        signal a workflow execution
     terminate, term  terminate a new workflow execution
     list             list open or closed workflow executions
     query            query workflow execution

OPTIONS:
   --help, -h  show help

```
the usage will indicate how the help option can be used

File: tools/cli/README.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+# for workflow with multiple input, seperate each json with space/newline like
+./cadence-cli --domain samples-domain workflow start --tl helloWorldGroup --wt main.WorkflowWith3Args --et 60 --dt 10 -i '"your_input_string" 123 {"Name":"my-string", "Age":12345}'
+```
```
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: Unfortunately we cannot. 
Because using your proposal there is no way to tell if `["your_input_string", 123, {"Name":"my-string", "Age":12345}]` input is 1 list argument or 3 arguments.

File: tools/cli/README.md (hunk 8 above)
Comment (vancexu) [reply tools/cli/README.md]: changed

File: tools/cli/README.md (hunk 9 above)
Comment (vancexu) [reply tools/cli/README.md]: With the cli framework we currently use, it would be a little bit tricky to achieve this requirement. Because `show` commands also accepts other optional arguments. 
Instead, I propose to add new command like `showid <wid>` as to separate command which can only take wid and have no other arguments to achieve this. what do you think?

File: tools/cli/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: Yes, added

File: tools/cli/factory.go (hunk 13 above)
Comment (vancexu) [reply tools/cli/factory.go]: Changed, I missed one part previously. Thanks

File: tools/cli/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: Renamed to cadence

[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: Done

File: tools/cli/README.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+# then just run commands without --domain flag, like
+./cadence domain desc
+```
+
+### Workflow operation examples
+(The following examples assume you already export CADENCE_CLI_DOMAIN environment variable as Tips above)
+
+- Run workflow: Start a workflow and see it's progress 
+```
+./cadence workflow run --tl helloWorldGroup --wt main.Workflow --et 60 --dt 10 -i '"cadence"'
```
Comment (mfateev) [inline tools/cli/README.md]: Let's default decision task timeout to something like 5 seconds and do not require it.

File: tools/cli/README.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+export CADENCE_CLI_DOMAIN=samples-domain
+
+# then just run commands without --domain flag, like
+./cadence domain desc
+```
+
+### Workflow operation examples
+(The following examples assume you already export CADENCE_CLI_DOMAIN environment variable as Tips above)
+
+- Run workflow: Start a workflow and see it's progress
```
Comment (mfateev) [inline tools/cli/README.md]: Add that this command doesn't finish until workflow completes.

File: tools/cli/README.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+./cadence workflow show -w 3ea6b242-b23c-4279-bb13-f215661b4717 -r 866ae14c-88cf-4f1e-980f-571e031d71b0
+
+# if run_id is not provided, it will show the latest run history for that workflow_id
+./cadence workflow show -w 3ea6b242-b23c-4279-bb13-f215661b4717
+```
+
+- List open or closed workflow executions
+```
+./cadence workflow list
+```
```
Comment (mfateev) [inline tools/cli/README.md]: What about query?

File: tools/cli/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (mfateev) [reply tools/cli/README.md]: We could say that list always defaults to the list of arguments. If they indeed want to specify list as a single argument (which is very rare) then they have to pass list of lists:
[[...]].

File: cmd/tools/cli/main.go [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+// Start using this CLI tool with command
+// `./cadence-cli` to show help message
+// `./cadence-cli domain` to show help message on domain operations
+// `./cadence-cli workflow` to show help message on workflow operations
+func main() {
+	app := cli.NewCliApp()
+
+	logger, err := zap.NewDevelopment()
```
Comment (yiminc) [inline cmd/tools/cli/main.go]: Can we not make this logger a required argument to NewBuilder()? It could create a default inside that builder. Just want to keep code here to be minimal.

File: cmd/tools/cli/main.go [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+	if err != nil {
+		panic(err)
+	}
+	cli.SetBuilder(cli.NewBuilder(logger))
```
Comment (yiminc) [inline cmd/tools/cli/main.go]: I would even make this call optional and use cli.NewBuilder() if it is not set.
Actually, I would check the builder inside the cli.NewCliApp() and set it to the default value if it is not set yet. So in internal version, you can call the cli.SetBuilder() before cli.NewCliApp().

File: tools/cli/README.md (hunk 16 above)
Comment (vancexu) [reply tools/cli/README.md]: Done, default 10 sec

File: tools/cli/README.md (hunk 17 above)
Comment (vancexu) [reply tools/cli/README.md]: Added

File: tools/cli/README.md (hunk 18 above)
Comment (vancexu) [reply tools/cli/README.md]: Added. Need to add samples to make it looks better later

File: cmd/tools/cli/main.go [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"os"
+
+	"github.com/uber/cadence/tools/cli"
+	"go.uber.org/zap"
+)
+
+// Start using this CLI tool with command
+// `./cadence-cli` to show help message
```
[the remark this answers is not in the document]
Comment (vancexu) [reply cmd/tools/cli/main.go]: updated

File: tools/cli/factory.go [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+	BuildServiceClient(c *cli.Context) (workflowserviceclient.Interface, error)
+}
+
+// WorkflowClientBuilder build client to cadence service
+type WorkflowClientBuilder struct {
+	hostPort   string
+	dispatcher *yarpc.Dispatcher
+	Logger     *zap.Logger
```
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/factory.go]: Good catch, changed Logger to logger. 
WorkflowClientBuilder is needed, otherwise lint complain on NewBuilder function.

File: cmd/tools/cli/main.go (hunk 19 above)
Comment (vancexu) [reply cmd/tools/cli/main.go]: Updated

File: cmd/tools/cli/main.go (hunk 20 above)
Comment (vancexu) [reply cmd/tools/cli/main.go]: Make sense I like it. Updated.

File: tools/cli/README.md [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+
+- Query workflow execution
+```
+# use build-in query type "__stack_trace" which is supported by cadence client library
+./cadence workflow query -w <wid> -r <rid>
```
Comment (mfateev) [inline tools/cli/README.md]: What about 
cadence workflow stack -w ...

File: tools/cli/README.md [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+
+Example uses [this cadence-samples workflow](https://github.com/samarabbas/cadence-samples/blob/master/cmd/samples/recipes/helloworld/helloworld_workflow.go) 
+and it takes a string as input, so there is the `-i '"cadence"'`. Single quote `''` is used to wrap input as json. 
+
+**Note:** you need to start worker so that workflow can make progress.  
+(Run `make && ./bin/helloworld -m worker` in cadence-samples to start the worker)
+
+- Start workflow: 
+```
+./cadence workflow start --tl helloWorldGroup --wt main.Workflow --et 60 --dt 10 -i '"cadence"'
```
Comment (mfateev) [inline tools/cli/README.md]: I would change ./cadence to cadence in all these samples. In most cases it would be installed though brew or be somewhere in the PATH.

File: tools/cli/README.md [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+- Query workflow execution
+```
+# use build-in query type "__stack_trace" which is supported by cadence client library
+./cadence workflow query -w <wid> -r <rid> 
+
+# use custom query type
+./cadence workflow query -w <wid> -r <rid> --qt <query-type>
+```
+
+- Signal, cancel, terminate workflow
```
Comment (mfateev) [inline tools/cli/README.md]: We have the open activities visibility query. Where is it?
Also we have list of workers connected to a task list. Where is it?

File: tools/cli/README.md (hunk 23 above)
Comment (vancexu) [reply tools/cli/README.md]: I think query at least link me with QueryWorkflow of cadence-client, no sure why you propose stack?

File: tools/cli/README.md (hunk 24 above)
Comment (vancexu) [reply tools/cli/README.md]: For this readme, I think `./cadence` is better because reader can easily understand it is the executable after running `make`, and "advanced player" can obviously use PATH and replace it to whatever they what. 
Directly put `cadence` instead of `./cadence` will confuse beginner, if I don't have instructions on how to set PATH or use brew. And for the simplicity of this readme, I don't want to include those.

File: tools/cli/README.md (hunk 25 above)
Comment (vancexu) [reply tools/cli/README.md]: Create https://github.com/uber/cadence/issues/588 and will work on separate pull request.

File: tools/cli/README.md (hunk 23 above)
Comment (mfateev) [reply tools/cli/README.md]: Because if you don't specify query type you default to stack which is confusing. I would require query type, but provide separate command "stack" which would execute query type "__stack_trace" which is hard to remember.

File: tools/cli/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (mfateev) [reply tools/cli/README.md]: Why is it resolved?

File: tools/cli/README.md (hunk 9 above)
Comment (mfateev) [reply tools/cli/README.md]: If it is indeed that hard then showid is OK.

File: tools/cli/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (vancexu) [reply tools/cli/README.md]: After discussion, we decide to stick with using spaces/newlines separate json, instead of the proposed using list.

File: tools/cli/README.md (hunk 9 above)
Comment (vancexu) [reply tools/cli/README.md]: Cool, added `showid <wid>` as a shortcut of `show -w <wid>`

File: tools/cli/README.md (hunk 23 above)
Comment (vancexu) [reply tools/cli/README.md]: Added `stack` command
